packages/shared/useFullscreen.ts

```diff
@@ -83,7 +83,11 @@ type FullscreenApi = {
  * useFullscreen
  * A hook that helps make the document fullscreen
  */
-function useFullscreen():FullscreenApi{
+function useFullscreen(): FullscreenApi | undefined {
+  if (typeof window === 'undefined') {
+    return;
```

Comment:
I think it'd be nice to return a negative match here rather than `undefined` to avoid users having to handle the SSR case specially e.g.

```
{
  isEnabled: false,
  isFullscreen: false,
  // ... etc
}
```

Reply:
Hmm I get your point, but not sure how would that be beneficial for the user? Will still need to do a check what was returned from the hook. With returning `undefined` it's clearly stated - "Hey, it's SSR, it won't work on the server". I'm happy to discuss more though :) What's your opinion on this @imbhargav5 ?

P.S. It definitely works with Next.js, and from my experience with SSR it will trigger `setState` when coming to the browser.

Reply:
Running tests now @maciekgrzybek . Going to nail this in the next 2 hours. This hook is by far the most insecure hook we ever made :D

Reply:
Yeah. I think it's best we return some falsy values on the server. Will do that now.

Reply:
cc @gpoole

Reply:
> not sure how would that be beneficial for the user?
The benefit I see would be that during SSR the user's code can follow the same path as when fullscreen support is not present, rather than needing to write a special case just for SSR.

> from my experience with SSR it will trigger setState when coming to the browser
The issue I'm thinking of would be that to work correctly the 2-pass steps in the client should be: render once with fullscreen support off (matching the server state), then render a second time if fullscreen support is detected (2nd pass). If you skip the first step immediately render in the fullscreen supported state, which is what I think is happening here, you may have a mismatch with the SSR output during the first hydration pass and could run into issues.

Reply:
Yeah. We can fail silently on the server as a convenience with falsy values but doing it in 2 steps is needlessly complicated for the client side only scenario.

Reply:
Also, as I am skimming through the code, I find the `onChange` and `onError` functions pretty difficult to work with and to refactor.

Reply:
Yeah I understand. The problem will be that not doing both passes will cause DOM mismatch problems when using SSR if different output is returned depending on whether fullscreen is available. It definitely is an unnecessary step for the client only scenario unfortunately, yes, but I'm not aware of a better way that avoids the issue.

Reply:
One thing I can think of is that it is an additional performance hit but also one that I would have thought was relatively minor for the majority of users. Potentially a flag to disable 2-pass for users concerned about performance could cover that?

Reply:
It's shaping out well now. Should be able to work solidly from now on.

Reply:
@gpoole Because useFullscreen has no purpose on the server, we shouldn't do anything more than silently failing on the server and optimise best for the client side where it is meant to be used. 
I feel like a lot of devs can simply use `next/dynamic` or `react-no-ssr` or something quite easily and wrap a component around and make it work client only for fullscreen. 
So it's best we make this hook work flawlessly on client and make it not too bad on server.

Reply:
👍 understood, that's fair enough